src/main/java/de/tum/in/www1/artemis/repository/ProgrammingSubmissionRepository.java

```diff
@@ -62,6 +63,20 @@ default ProgrammingSubmission findFirstByParticipationIdAndCommitHashOrderByIdDe
             """)
     Optional<ProgrammingSubmission> findFirstByParticipationIdOrderByLegalSubmissionDateDesc(@Param("participationId") Long participationId);
 
+    /**
+     * Get the latest legal submission for a participation (ignoring illegal submissions).
+     * The type of the submission can also be null, so that the query also returns submissions without a type.
+     *
+     * @param participationId the id of the participation
+     * @return the latest legal submission for the participation
+     */
+    default Optional<ProgrammingSubmission> findFirstByParticipationIdWithResultsOrderByLegalSubmissionDateDesc(Long participationId) {
```

Comment:
when those queries are only used in tests, I suggest to move them into a test repository